### PR TITLE
Implement built-in threading

### PR DESCRIPTION
**Pull request by philippjfr**
Allows declaring `pn.config.nthreads` and ``--num-threads`` on the CLI which launches a `ThreadPoolExecutor` to which we dispatch tasks as events arrive from the frontend.

### CODE REVIEW

**Comment by MarcSkovMadsen** [comment]
You might find inspiration in the example provided at https://github.com/holoviz/panel/issues/2631#issuecomment-898169379.

It's a different approach where you can submit `Job`s to the queue of a `BackgroundWorker` . My requirements for something like that is easy to use

1. Logging to console and `Terminal`.
2. Progress reporting
    - Number of jobs currently in queue
    - Index and instance of current job
3. Easy way to set `.loading` to True for a list of Panels for a specific `Job` or a `BackgroundWorker` while its running.

**Comment by philippjfr** [comment]
@MarcSkovMadsen @jbednar @maximlt Please review.

File: examples/user_guide/Performance_and_Debugging.ipynb [hunk 1]
```diff
@@ -46,6 +46,71 @@
...
+    "    async with aiohttp.ClientSession() as session:\n",
+    "        async with session.get(f\"https://picsum.photos/800/300?image={event.new}\") as resp:\n",
+    "            image.object = await resp.read()\n",
+    "            \n",
+    "widget.param.watch(update_img, 'value')\n",
+    "widget.param.trigger('value')\n",
+    "            \n",
+    "pn.Column(widget, image)"
+   ]
+  },
```
Comment (jbednar) [inline examples/user_guide/Performance_and_Debugging.ipynb:113]: Again, explicitly stating what the sequence of triggers and events would be here will help people understand how it works.

File: examples/user_guide/Deploy_and_Export.ipynb [hunk 2]
```diff
@@ -382,6 +382,7 @@
     "                            no index is served.\n",
     "      --num-procs N         Number of worker processes for an app. Using 0 will\n",
     "                            autodetect number of cores (defaults to 1)\n",
+    "      --num-threads N       Number of threads to launch in a ThreadPoolExecutor which Panel will dispatch events to. \n",
```
[the remark this answers is not in the document]
Comment (philippjfr) [reply examples/user_guide/Deploy_and_Export.ipynb:385]: Default is no thread pool, will clarify.

File: examples/user_guide/Performance_and_Debugging.ipynb [hunk 3]
```diff
@@ -46,6 +46,71 @@
...
   },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Threading and Async IO\n",
+    "\n",
+    "Threading and async are both approaches to speed up processing in Python using concurrency. Since we can't provide a complete primer on either threading or async here, if you are not familiar with these concepts we recommend reading up on them before continuing. Read about [threading in Python here](https://realpython.com/intro-to-python-threading/) and [AsyncIO here](https://realpython.com/async-io-python/).\n",
+    "\n",
+    "When to use which approach cannot be answered easily however as a general guide use `asyncio` when you are performing many slow IO bound operations simultaneously and use threading when you have a limited amount of connections or file operations. In some situations threading can also be useful for CPU intensive operations where the operation being executed [releases the GIL](https://realpython.com/async-io-python/), this includes many NumPy, Pandas and Numba functions.\n",
```
[the remark this answers is not in the document]
Comment (philippjfr) [reply examples/user_guide/Performance_and_Debugging.ipynb:57]: I mean it's not really a clear cut thing, but basically asyncio will work well if you're making 100 or 1000s of requests or file I/O operations, while threading is best for a much smaller number (i.e. limited by the number of threads available). Not sure how best to clarify phrase that tbh.

File: examples/user_guide/Performance_and_Debugging.ipynb [hunk 4]
```diff
@@ -46,6 +46,71 @@
...
+    "widget = pn.widgets.IntSlider(start=0, end=10)\n",
+    "\n",
+    "async def get_img(index):\n",
+    "    async with aiohttp.ClientSession() as session:\n",
+    "        async with session.get(f\"https://picsum.photos/800/300?image={index}\") as resp:\n",
+    "            return pn.pane.JPG(await resp.read())\n",
+    "            \n",
+    "pn.Column(widget, pn.bind(get_img, widget))"
+   ]
+  },
```
[the remark this answers is not in the document]
Comment (philippjfr) [reply examples/user_guide/Performance_and_Debugging.ipynb:85]: That's right, will clarify!

File: examples/user_guide/Performance_and_Debugging.ipynb (hunk 3 above)
[the remark this answers is not in the document]
Comment (jbednar) [reply examples/user_guide/Performance_and_Debugging.ipynb:57]: It's not clear to me whether the two options are mutually exclusive, but in any case, just saying explicitly what you say here would help.

File: examples/user_guide/Performance_and_Debugging.ipynb [hunk 5]
```diff
@@ -46,6 +46,80 @@
...
+    "import aiohttp\n",
+    "\n",
+    "widget = pn.widgets.IntSlider(start=0, end=10)\n",
+    "\n",
+    "async def get_img(index):\n",
+    "    async with aiohttp.ClientSession() as session:\n",
+    "        async with session.get(f\"https://picsum.photos/800/300?image={index}\") as resp:\n",
+    "            return pn.pane.JPG(await resp.read())\n",
+    "            \n",
+    "pn.Column(widget, pn.bind(get_img, widget))"
```
Comment (maximlt) [inline examples/user_guide/Performance_and_Debugging.ipynb:83]: I could not get this example to work, either in a notebook or as a script. Running the script showed this warning:
```
RuntimeWarning: coroutine 'get_img' was never awaited
  return f(self, obj, val)
```
I used python 3.8.

File: examples/user_guide/Performance_and_Debugging.ipynb [hunk 6]
```diff
@@ -46,6 +46,80 @@
...
+    "\n",
+    "async def update_img(event):\n",
+    "    async with aiohttp.ClientSession() as session:\n",
+    "        async with session.get(f\"https://picsum.photos/800/300?image={event.new}\") as resp:\n",
+    "            image.object = await resp.read()\n",
+    "            \n",
+    "widget.param.watch(update_img, 'value')\n",
+    "widget.param.trigger('value')\n",
+    "            \n",
+    "pn.Column(widget, image)"
```
Comment (maximlt) [inline examples/user_guide/Performance_and_Debugging.ipynb:113]: Also unable to run this snippet either in a notebook or in a script.

When I launch the app from a script I get the following error:
```
2021-12-01 18:04:04,067 Exception in callback functools.partial(<bound method IOLoop._discard_future_result of <tornado.platform.asyncio.AsyncIOMainLoop object at 0x7fa69897b250>>, <Task finished name='Task-8' coro=<_needs_document_lock.<locals>._needs_document_lock_wrapper() done, defined at /Users/mliquet/miniconda3/envs/panel-dev38/lib/python3.8/site-packages/bokeh/server/session.py:77> exception=TypeError("argument should be a bytes-like object or ASCII string, not 'NoneType'")>)
Traceback (most recent call last):
  File "/Users/mliquet/miniconda3/envs/panel-dev38/lib/python3.8/site-packages/tornado/ioloop.py", line 741, in _run_callback
    ret = callback()
  File "/Users/mliquet/miniconda3/envs/panel-dev38/lib/python3.8/site-packages/tornado/ioloop.py", line 765, in _discard_future_result
    future.result()
  File "/Users/mliquet/miniconda3/envs/panel-dev38/lib/python3.8/site-packages/bokeh/server/session.py", line 93, in _needs_document_lock_wrapper
    result = func(self, *args, **kwargs)
  File "/Users/mliquet/miniconda3/envs/panel-dev38/lib/python3.8/site-packages/bokeh/server/session.py", line 227, in with_document_locked
    return func(*args, **kwargs)
  File "/Users/mliquet/miniconda3/envs/panel-dev38/lib/python3.8/site-packages/bokeh/document/callbacks.py", line 450, in wrapper
    return invoke_with_curdoc(doc, invoke)
  File "/Users/mliquet/miniconda3/envs/panel-dev38/lib/python3.8/site-packages/bokeh/document/callbacks.py", line 408, in invoke_with_curdoc
    return f()
  File "/Users/mliquet/miniconda3/envs/panel-dev38/lib/python3.8/site-packages/bokeh/document/callbacks.py", line 449, in invoke
    return f(*args, **kwargs)
  File "/Users/mliquet/miniconda3/envs/panel-dev38/lib/python3.8/site-packages/bokeh/document/callbacks.py", line 179, in remove_then_invoke
    return callback()
  File "/Users/mliquet/miniconda3/envs/panel-dev38/lib/python3.8/site-packages/panel/pane/base.py", line 153, in _update_object
    self._update(ref, old_model)
  File "/Users/mliquet/miniconda3/envs/panel-dev38/lib/python3.8/site-packages/panel/pane/markup.py", line 48, in _update
    model.update(**self._get_properties())
  File "/Users/mliquet/miniconda3/envs/panel-dev38/lib/python3.8/site-packages/panel/pane/image.py", line 117, in _get_properties
    data = base64.b64decode(data)
  File "/Users/mliquet/miniconda3/envs/panel-dev38/lib/python3.8/base64.py", line 80, in b64decode
    s = _bytes_from_decode_data(s)
  File "/Users/mliquet/miniconda3/envs/panel-dev38/lib/python3.8/base64.py", line 45, in _bytes_from_decode_data
    raise TypeError("argument should be a bytes-like object or ASCII "
TypeError: argument should be a bytes-like object or ASCII string, not 'NoneType'
```

**Review by maximlt — COMMENTED** [review]
This is going in a super nice direction!

I'd suggest to expand the documentation a little, trying to make it clearer what this is all about to people who are not software developers, which I assume (might be wrong!) that it's the majority of Panel users.
* IO bounds vs CPU bounds operations?
* Provide an example for Threading. Two actually ideally, the second one showing how to take advantage of releasing the GIL.
* Are the aysncIO examples actually relevant here? The doc starts by stating that asyncIO is useful in the order of 100 or more operations.
* Does the usage of threading/asyncio mean that an application scale better (i.e. is less affected by an increasing number of users)?
* The Threading section says that whenever an event is generated Panel will dispatch it. It'd be nice to describe in more details what an *event* is here.

File: examples/user_guide/Performance_and_Debugging.ipynb (hunk 6 above)
Comment (philippjfr) [reply examples/user_guide/Performance_and_Debugging.ipynb:113]: Requires master, which version are you running?

File: examples/user_guide/Performance_and_Debugging.ipynb (hunk 5 above)
Comment (philippjfr) [reply examples/user_guide/Performance_and_Debugging.ipynb:83]: Again, requires master.

File: examples/user_guide/Performance_and_Debugging.ipynb [hunk 7]
```diff
@@ -46,6 +46,80 @@
...
+    "When using Python>=3.8 you can use async callbacks wherever you would ordinarily use a regular synchronous function. For instance you can use `pn.bind` on an async function:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import aiohttp\n",
```
[the remark this answers is not in the document]
Comment (philippjfr) [reply examples/user_guide/Performance_and_Debugging.ipynb:74]: Thanks, good catch!

File: examples/user_guide/Performance_and_Debugging.ipynb [hunk 8]
```diff
@@ -46,6 +46,80 @@
...
   },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Threading and Async IO\n",
+    "\n",
+    "Threading and async are both approaches to speed up processing in Python using concurrency. Since we can't provide a complete primer on either threading or asynchronous processing here, if you are not familiar with these concepts we recommend reading up on them before continuing. Read about [threading in Python here](https://realpython.com/intro-to-python-threading/) and [AsyncIO here](https://realpython.com/async-io-python/).\n",
+    "\n",
+    "When to use which approach cannot be answered easily and is never completely clear cut. As a general guide however use `asyncio` when you are performing many slow IO bound operations concurrently (on the order of 100 and more) and use threading when you are performing a limited number (less than the number of available threads). In some situations threading can also be useful for CPU intensive operations where the operation being executed [releases the GIL](https://realpython.com/async-io-python/), this includes many NumPy, Pandas and Numba functions.\n",
```
[the remark this answers is not in the document]
Comment (philippjfr) [reply examples/user_guide/Performance_and_Debugging.ipynb:57]: Good catch!

File: examples/user_guide/Performance_and_Debugging.ipynb (hunk 6 above)
Comment (maximlt) [reply examples/user_guide/Performance_and_Debugging.ipynb:113]: I run them from this branch, `enable_threading`.

Comment (philippjfr) [reply examples/user_guide/Performance_and_Debugging.ipynb:113]: Guess it needs to be rebased.

File: examples/user_guide/Performance_and_Debugging.ipynb [hunk 9]
```diff
@@ -46,6 +46,80 @@
...
+    "\n",
+    "### Threading\n",
+    "\n",
+    "Threading in Panel can be enabled easily by setting the `config.nthreads` parameter. This will start a `ThreadPoolExecutor` with the specified number of threads (or if set to `0` it will set the number of threads based on your system, i.e. `min(32, os.cpu_count() + 4)`). Whenever an event is generated or a periodic callback fires Panel will then automatically dispatch the event to the executor. If launching an application with `panel serve` you can configure this option with the `--num-threads` option.\n",
```
[the remark this answers is not in the document]
Comment (philippjfr) [reply examples/user_guide/Performance_and_Debugging.ipynb:61]: Hmm, odd, both should work. Will investigate.

**Pull request merged**